feature/session/src/main/java/io/github/droidkaigi/confsched2019/session/ui/widget/SessionToolbarBehavior.kt

```diff
@@ -0,0 +1,131 @@
+package io.github.droidkaigi.confsched2019.session.ui.widget
+
+import android.content.Context
+import android.util.AttributeSet
+import android.util.TypedValue
+import android.view.View
+import android.view.animation.AccelerateDecelerateInterpolator
+import android.widget.LinearLayout
+import androidx.appcompat.widget.AppCompatTextView
+import androidx.appcompat.widget.Toolbar
+import androidx.coordinatorlayout.widget.CoordinatorLayout
+import androidx.core.view.ViewCompat
+import androidx.core.view.ViewPropertyAnimatorListener
+import androidx.core.view.children
+import androidx.core.widget.NestedScrollView
+import io.github.droidkaigi.confsched2019.session.R
+
+class SessionToolbarBehavior(
+    val context: Context? = null,
+    attrs: AttributeSet? = null
+) : CoordinatorLayout.Behavior<LinearLayout>(context, attrs) {
+
+    private var sessionTitle: String = ""
+    private var hasSetToolbarTitle = false
+    private var appCompatTextView: AppCompatTextView? = null
+
+    private var mIsAnimation = false
+
+    constructor(context: Context?, attrs: AttributeSet?, sessionTitle: String) : this(
+        context,
+        attrs
+    ) {
+        this.sessionTitle = sessionTitle
+    }
+
+    override fun onStartNestedScroll(
+        coordinatorLayout: CoordinatorLayout,
+        child: LinearLayout,
+        directTargetChild: View,
+        target: View,
+        axes: Int,
+        type: Int
+    ): Boolean = axes == ViewCompat.SCROLL_AXIS_VERTICAL
+
+    override fun onNestedPreScroll(
+        coordinatorLayout: CoordinatorLayout,
+        child: LinearLayout,
+        target: View,
+        dx: Int,
+        dy: Int,
+        consumed: IntArray,
+        type: Int
+    ) {
+        val toolbar = child.findViewById<Toolbar>(R.id.session_toolbar)
+        val nestedScrollView = if (child.childCount > 0) {
+            (child.children.find { it is NestedScrollView } as? NestedScrollView)
+        } else {
+            null
+        }
+        performAnimation(toolbar, nestedScrollView)
+    }
+
+    private fun performAnimation(toolbar: Toolbar, nestedScrollView: NestedScrollView?) {
+        if (!mIsAnimation) {
+            if (nestedScrollView?.canScrollVertically(NEGATIVE_DIRECTION) == false) {
+                appCompatTextView?.let {
+                    animateTitle(TOP_TITLE_ANIMATION_ALPHA, TOP_TITLE_ANIMATION_DURATION, it)
+                    animateToolbarElevation(toolbar, TOP_TOOLBAR_ELEVATION)
+                }
+            } else {
+                initToolbarTitle(toolbar)
+                appCompatTextView?.let {
+                    animateTitle(NO_TOP_TITLE_ANIMATION_ALPHA, NO_TOP_TITLE_ANIMATION_DURATION, it)
+                    animateToolbarElevation(toolbar, NO_TOP_TOOLBAR_ELEVATION)
+                }
+            }
+        }
+    }
+
+    private fun initToolbarTitle(toolbar: Toolbar) {
+        if (!hasSetToolbarTitle) {
+            toolbar.title = sessionTitle
+            appCompatTextView =
```

Comment:
I guess this cause app crash if any view is expanded while processing this logic. For detail, please have a look at `Toolbar#setText` and `Toolbar#setSystemView`.